### DISCUSSION TITLE
Announcing TriliumAlchemy: a Python SDK and CLI toolkit for Trilium

### DISCUSSION THREAD

**Discussion by mm21** [TriliumNext/Trilium#4144] [Show and tell]
The project is still in an early stage, but should be stable. All feedback is welcomed.

Repo: https://github.com/mm21/trilium-alchemy/
Docs: https://mm21.github.io/trilium-alchemy/

Tagging folks who expressed interest before: @Nriver @perfectra1n @BeatLink @soulsands @zadam @jwhonce

**Comment by Nriver** [comment]
Nice work!

**Reply by Nriver** [reply]
added to https://github.com/Nriver/awesome-trilium

**Reply by mm21** [reply]
Thanks! This is a really nice list, I refer to it often.

**Comment by meichthys** [comment]
@mm21 Do you know if this would build for arm/android?

**Reply by mm21** [reply]
Great question, I haven't tried but will add this to the roadmap. I don't see why it wouldn't as the dependencies are minimal.

**Comment by jwhonce** [comment]
@mm21 An imperative example, https://github.com/jwhonce/trilium-addons/blob/main/TaskManager/tm.py

**Reply by mm21** [reply]
Very cool! This is exactly what I was hoping it would be used for. I've been meaning to try Typer for my next CLI project, it looks really nice.

Just had a quick comment about the usage - passing `leaf=True` in `Note()` isn't needed, actually it doesn't take a `leaf` argument. I'll add a warning for unexpected kwargs.

For more background, `leaf` is only relevant for `Note` subclasses which are singletons, i.e. they have a deterministic `note_id`, or `note_id` was passed during instantiation. Children of `Note` subclasses are expected to be managed by `@children` or `@child` decorators, unless `leaf` is set, in which case they're managed by the user. For `Note` itself, it won't tamper with any child notes unless you explicitly pass `children` to set them.

Nice to see the `template` arg of `Note()` being used, that was a last minute feature.

**Reply by jwhonce** [reply]
Thanks for review and feedback, I will make changes. 

Working with attributes was so much easier with the abstraction vs. `ETAPI`. I was able to delete a lot of housekeeping code. One enhancement I think I would like, is some property/method to determine from my code if an attribute is multiple vs single. It seems smelly that I have to know that detail.

I have been playing with `Typer` for these scripts. So far I like it. It is a bit overkill for this script but this script did made a good learning example. Having autocompletion on the Task descriptions is just plain cool. :+1: The `rich` library is also very nice to work with for printing tables.

**Reply by mm21** [reply]
Glad to hear it was easier to work with attributes. 

Good idea to have a way to check if an attribute is multiple vs single. You mean a mechanism to get the attribute definition (e.g. `#label:myLabel=promoted,single,text`) right? Currently you can use `len(note.attributes["myLabel"])` to get the number of attributes by name, but that doesn't tell you if it's multi- or single-valued according to the attribute definition.

I've proposed a feature for further discussion here: https://github.com/mm21/trilium-alchemy/issues/3
